### PR TITLE
ref: SentryUIPerformanceTracker

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Separated the UIPerformanceTracker behaviour from swizzling.
_#skip-changelog_
## :bulb: Motivation and Context

Easier to test and maintain.

## :green_heart: How did you test it?

Unit tests

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [ ] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

File: Tests/SentryTests/Integrations/SentryUIPerformanceTrackerTests.swift [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        var date = Date(timeIntervalSince1970: 0)
+        
+        fixture.dateProvider.setDate(date: date)
+        var lastSpan: Span?
+        
+        sut.viewControllerLoadView(viewController) {
+            transactionSpan = self.getStack(tracker: tracker).first
+            lastSpan = self.getStack(tracker: tracker).last!
+            date = date.addingTimeInterval(1)
+            self.fixture.dateProvider.setDate(date: date)
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/SentryUIPerformanceTrackerTests.swift:116]: `l`: You could extract these two into a function and call it for example `advanceTimeBy`. I use this approach in many tests, see 
https://github.com/getsentry/sentry-cocoa/blob/ef7ba5fedd7afc40710567fdb7427375a3d52f29/Tests/SentryTests/Integrations/SentrySessionTrackerTests.swift#L331-L346

File: Tests/SentryTests/Integrations/SentryUIPerformanceTrackerTests.swift [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+            
+        }
+        
+        XCTAssertFalse(transactionSpan.isFinished)
+        tracker.finishSpan(customSpanId!)
+        XCTAssertTrue(transactionSpan.isFinished)
+        XCTAssertEqual(Dynamic(transactionSpan).children.asArray!.count, 3)
+    }
+    
+    func testSkipLoadView() {
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/SentryUIPerformanceTrackerTests.swift:189]: I like testing this edge case.

File: Tests/SentryTests/Integrations/SentryUIPerformanceTrackerTests.swift [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            lastSpan = self.getStack(tracker: tracker).last!
+            date = date.addingTimeInterval(5)
+            self.fixture.dateProvider.setDate(date: date)
+        }
+        assertSpanDuration(span: lastSpan!, expectedDuration: 5)
+        
+        assertSpanDuration(span: transactionSpan, expectedDuration: 18)
+    }
+    
+    func testWaitingForCustomSpan() {
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/SentryUIPerformanceTrackerTests.swift:163]: Nice test case.

**Review by philipphofmann — COMMENTED** [review]
Thanks for the refactoring. This makes the code way better IMO. I added a few comments to hopefully even improve it a bit.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,9 @@
 
 ## 7.2.0-beta.0
 
+- fix: Set description for performance tracker spans (#1042)
+- fix: Capturing Child Spans (#1139)
+- feat: Auto UI Performance Instrumentation (#1105)
```
[the remark this answers is not in the document]
Comment (brustolin) [reply CHANGELOG.md:7]: This is odd, I did not update changelog, this should be something left from merge.

File: Sources/Sentry/Public/SentryDefines.h [hunk 5]
```diff
@@ -25,6 +25,11 @@
 @class SentryEvent, SentryBreadcrumb, SentrySamplingContext;
 @protocol SentrySpan;
 
+/**
+ * Block for general purpose callback
+ */
+typedef void (^Callback)(void);
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/Public/SentryDefines.h:31]: I don't have a strong opinion about having this typedef, but I do believe that's no reason to make it public.
I will remove it.

File: Sources/Sentry/SentryUISwizziling.m [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+#import "SentrySwizzle.h"
+#import "SentryUIPerformanceTracker.h"
+#import "SentryUIViewControllerSanitizer.h"
+#import <objc/runtime.h>
+
+#if SENTRY_HAS_UIKIT
+#    import <UIKit/UIKit.h>
+#endif
+
+@implementation SentryUISwizziling
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/SentryUISwizziling.m:17]: I was thinking about the future where we could have more swizzling here, maybe something from SwiftUI.

File: Sources/Sentry/SentryUISwizziling.m [hunk 7]
```diff
@@ -0,0 +1,156 @@
+#import "SentryUISwizziling.h"
+#import "SentryHub.h"
+#import "SentryLog.h"
+#import "SentryPerformanceTracker.h"
+#import "SentrySDK+Private.h"
+#import "SentryScope.h"
+#import "SentrySpanId.h"
+#import "SentrySwizzle.h"
+#import "SentryUIPerformanceTracker.h"
+#import "SentryUIViewControllerSanitizer.h"
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/SentryUISwizziling.m:10]: Definitely. Will take a look.

File: Sources/Sentry/SentryUIPerformanceTracker.m [hunk 8]
```diff
@@ -9,229 +9,134 @@
...
-    if (appImage == nil) {
-        if ([UIApplication respondsToSelector:@selector(sharedApplication)]) {
-            UIApplication *app = [UIApplication performSelector:@selector(sharedApplication)];
-            appImage = class_getImageName(app.delegate.class);
-        }
+    SentrySpanId *spanId = objc_getAssociatedObject(target, &SENTRY_UI_PERFORMANCE_TRACKER_SPAN_ID);
+
+    if (spanId == nil) {
+        callback();
+    } else {
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/SentryUIPerformanceTracker.m:51]: The most significant reason I can think of is if the user clears all associated objects from the view controller using ```objc_removeAssociatedObjects```

File: Sources/Sentry/SentryUIPerformanceTracker.m [hunk 9]
```diff
@@ -9,229 +9,134 @@
...
-            }
-        }),
-        SentrySwizzleModeOncePerClassAndSuperclasses, (void *)selector);
+    [self measurePerformance:@"viewDidAppear" target:controller blockToMeasure:callback];
+
+    SentrySpanId *spanId
+        = objc_getAssociatedObject(controller, &SENTRY_UI_PERFORMANCE_TRACKER_SPAN_ID);
+    if (spanId != nil) {
+        [self.tracker finishSpan:spanId];
+    }
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/SentryUIPerformanceTracker.m:92]: Same as https://github.com/getsentry/sentry-cocoa/pull/1160/files#r652240299

File: Sources/Sentry/include/SentryUIPerformanceTracker.h [hunk 10]
```diff
@@ -1,16 +1,91 @@
...
 
 /**
- * Class is responsible to swizzle UI key methods
- * so Sentry can track UI performance.
+ * Class responsible to track UI performance.
+ * This class is intended to be used in a swizzled context.
  */
 @interface SentryUIPerformanceTracker : NSObject
+#if SENTRY_HAS_UIKIT
+@property (nonatomic, readonly, class) SentryUIPerformanceTracker *shared;
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/include/SentryUIPerformanceTracker.h:18]: same for https://github.com/getsentry/sentry-cocoa/pull/1160/files#r651679406

File: Sources/Sentry/include/SentryUIPerformanceTracker.h [hunk 11]
```diff
@@ -1,16 +1,91 @@
...
+
+/**
+ * Measures viewController`s loadView method.
+ * This method starts a span that will be finished when
+ * viewControllerDidAppear:callBackToOrigin: is called.
+ *
+ * @param controller UIViewController to be measured
+ * @param callback A callback that indicates the swizzler to call the original view controller
+ * LoadView method.
+ */
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/include/SentryUIPerformanceTracker.h:28]: Right, I'll keep next documentation simpler for inner classes.

File: Sources/Sentry/SentryUISwizziling.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/SentryUISwizziling.m:11]: As the class is not public, we can then rename it. Now it's only about `UIViewControllers`, so I think it makes sense to rename it.

**Review by philipphofmann — APPROVED** [review]
LGTM thanks 🙏

**Pull request merged**